# OptiType Guide
This is a simple guide to install and run the HLA typing tool [OptiType](https://github.com/FRED-2/OptiType) from command line. 

Reference: [SZOLEK, et al. OptiType: precision HLA typing from next-generation sequencing data. 2014](https://pubmed.ncbi.nlm.nih.gov/25143287/).

## Dependencies
Despite OptiType utilizing a series of programs (Python, RazerS, SAMtools, HDF5, CPLEX) and Python modules (NumPy, Pyomo, PyTables, Pandas, Pysam, Matplotlib, Future) to function, the installation of each tool on your computer is not mandatory.


The tool is available via Docker, a platform that enables the packaging of an application or environment within a container, allowing anyone to run the program as long as Docker is installed on their computer. To install Docker, run the following the commands below:

```
sudo apt update
sudo apt install apt-transport-https ca-certificates curl software-properties-common
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu focal stable"
apt-cache policy docker-ce
sudo apt install docker-ce
```

In its [page](https://www.docker.com/) they offer executables to download and install docker in your machine. To install docker by this way, follow the [step-by-step installation guide](https://docs.docker.com/engine/install/) (Docker is available for all operating systems).

## Installing OptiType
With Docker installed, we will proceed with the installation of OptiType. To do this, simply use the command: ```docker pull fred2/optitype```.

## How to run OptiType
After downloading the Docker image, you can use the following command to run OptiType:

```
mkdir /path/to/your/directory/results/
chmod 777 /path/to/your/directory/results/
docker run -v /path/to/your/directory/:/data/ -t fred2/optitype -i input1 [input2] (-r|-d) -o /data/results/
```
The flags for OptiType are:

```
Verbose (-v)-  The path to the directory on your computer where the data to be HLA typed by OptiType is located should be directed to the /data/ directory within the Docker image using (:) (Example: /path/to/your/directory/:/data/)
Image tag (-t fred2/optitype)- Docker image of OptiType 
Input (-i)- .fastq file(s) or .bam file (previous generated by OptiType, and stored for reanalysis). One file: single-end mode; Two files: paired-end mode.
RNA/DNA (-r|-d)- Indicates type os sequencing: DNA or RNA 
Output (-o)- Where the analysis results will be generated. The output should be directed to the /data/results/ directory.
```

## Running OptiType test
To run the OptiType test, follow the step-by-step instructions below:

  1. Copy the OptiType GitHub directory to your computer: ```git clone https://github.com/FRED-2/OptiType.git```
  2. Copy the test files to your analysis directory:

```
## For DNA data test:
cp OptiType/test/exome/* /path/to/your/directory/

## For RNA data test:
cp OptiType/test/rna/* /path/to/your/directory/
```
  3. Create a "results" folder in your directory and grant global permissions to the folder.
```
##  Create a "results" folder:
mkdir /path/to/your/directory/results/

## Grant global permissions to the "results" directory
chmod 777 /path/to/your/directory/results/
```
  4. Run the following commands:
```
## Command for DNA data testing:
docker run --name test -v /path/to/your/directory/:/data/ -t fred2/optitype -i NA11995_SRR766010_1_fished.fastq NA11995_SRR766010_2_fished.fastq -d -o /data/results/ -v

## Command for RNA data testing:
docker run --name test -v /path/to/your/directory/:/data/ -t fred2/optitype -i CRC_81_N_1_fished.fastq CRC_81_N_2_fished.fastq -r -o /data/results/ -v
```
  5. To view the result, simply use the following command:

```
less -S /path/to/your/directory/results/data_hour/data_hour_result.tsv
```
The result should be as follows:

| A1	| A2	| B1	| B2	| C1	| C2	| Reads	| Objective |
| ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| A*01:01	| A*01:01	| B*08:01	| B*57:01	| C*07:01	| C*06:02	| 1156.0	| 1135.192 |


